pynav/imm.py

```diff
@@ -0,0 +1,474 @@
+        """Carries out prediction step for each model of the IMM. 
+
+        Parameters
+        ----------
+        x_km_models : List[StateWithCovariance]
+            List of model estimates from previous timestep, after mixing. 
+        u : StampedValue
+            Input
+        dt : Float
+            Timestep
+
+        Returns
+        -------
+        List[StateWithCovariance]
+            List of model estimates after prediction step is carried out. 
+        """
+        x_km_models = x_km_models.copy() 
+        x_check = []
+        for lv1, kf in enumerate(self.kf_list):
+            x_check.append(kf.predict(x_km_models[lv1], u, dt))
+        return x_check
+
+    def correct(self, x_models_check: List[StateWithCovariance], 
+                        mu_km_models: List[float], 
+                        y: Measurement, 
+                        u: StampedValue):
+        """Carry out the correction step for each model and update model probabilities. 
+
+        Parameters
+        ----------
+        x_models_check : List[StateWithCovariance]
+            Estimates from each model before correction.
+        mu_km_models : List[Float]
+            Probabilities for each model from previous timestep.
+        y : Measurement
+            Measurement to be fused into the current state estimate.
+        u: StampedValue
+            Most recent input, to be used to predict the state forward 
+            if the measurement stamp is larger than the state stamp.
+
+
+        Returns
+        -------
+        List[StateWithCovariance]
+            The corrected state estimates for each model
+        List[Float]
+            Updated model probabilities.
+        """
+        x_models_check = x_models_check.copy() 
+        mu_km_models = mu_km_models.copy()
+        N_MODES = len(x_models_check)
+        mu_k = np.zeros(N_MODES)
+
+        # Compute each model's normalization constant
+        c_bar = np.zeros(N_MODES)
+        for i in range(N_MODES):
+            for j in range(N_MODES):
+                c_bar[j] = c_bar[j]+self.Pi[i, j]*mu_km_models[i]
+
+        # Correct and update model probabilities
+        x_hat = []
+        for lv1, kf in enumerate(self.kf_list):
+            x, details_dict = \
+                kf.correct(x_models_check[lv1], y, u, output_details = True)
+            x_hat.append(x)
+            z = details_dict["z"]
+            S = details_dict["S"]
+            model_likelihood = multivariate_normal.pdf(z, mean=np.zeros(z.shape), cov=S)
+            mu_k[lv1] = model_likelihood*c_bar[lv1]
+        
+        # If all model likelihoods are zero to machine tolerance, np.sum(mu_k)=0 and it fails
+        # Add this fudge factor to get through those cases.
+        if np.allclose(mu_k, np.zeros(mu_k.shape)):
+            mu_k = 1e-10 * np.ones(mu_k.shape)
+
+        mu_k = mu_k/np.sum(mu_k)  
+
+        return x_hat, mu_k
+
+
+def run_time_varying_Q_filter(
```

Comment:
I vote to remove this and handle time-varying Q the more natural way with our framework

Reply:
I do handle the time-varying Q in our framework. Its just that to run a filter which assumes a time-varying Q it needs to be reinitialized at every step, which is not the case in current run_filter method

Reply:
Ah right just saw the time-varying example update. Yeah this will work.